src/ott/neural/flow_models/genot.py

```diff
@@ -0,0 +1,495 @@
+      kwargs: Keyword arguments for the ODE solver.
+
+    Returns:
+      The push-forward or pull-back distribution defined by the learnt
+      transport plan.
+
+    """
+    rng = utils.default_prng_key(rng)
+    if not forward:
+      raise NotImplementedError
+    if condition is not None:
+      assert len(source) == len(condition), (len(source), len(condition))
+    latent_batch = self.latent_noise_fn(rng, shape=(len(source),))
+    cond_input = source if condition is None else (
+        jnp.concatenate([source, condition], axis=-1)
+    )
+
+    @jax.jit
+    def solve_ode(input: jnp.ndarray, cond: jnp.ndarray) -> jnp.ndarray:
+      ode_term = diffrax.ODETerm(
+          lambda t, x, args: self.state_velocity_field.
+          apply_fn({"params": self.state_velocity_field.params},
+                   t=t,
+                   x=x,
+                   condition=cond)
+      ),
+      solver = kwargs.pop("solver", diffrax.Tsit5())
+      stepsize_controller = kwargs.pop(
+          "stepsize_controller", diffrax.PIDController(rtol=1e-5, atol=1e-5)
+      )
+      sol = diffrax.diffeqsolve(
+          ode_term,
+          solver,
+          t0=t_0,
+          t1=t_1,
+          dt0=kwargs.pop("dt0", None),
+          y0=input,
+          stepsize_controller=stepsize_controller,
+          **kwargs,
+      )
+      return sol.ys[0]
+
+    return jax.vmap(solve_ode)(latent_batch, cond_input)
+
+  def _valid_step(self, valid_loader, iter):
+    pass
+
+  @property
+  def learn_rescaling(self) -> bool:
+    """Whether to learn at least one rescaling factor."""
+    return (
+        self.unbalancedness_handler.rescaling_a is not None or
+        self.unbalancedness_handler.rescaling_b is not None
+    )
+
+  def _reshape_samples(self, arrays: Tuple[jnp.ndarray, ...],
+                       batch_size: int) -> Tuple[jnp.ndarray, ...]:
+    return jax.tree_util.tree_map(
+        lambda x: jnp.reshape(x, (batch_size * self.k_samples_per_x, -1)),
+        arrays
+    )
+
+  def _learn_rescaling(
+      self, source: jnp.ndarray, target: jnp.ndarray,
+      source_conditions: Optional[jnp.ndarray], tmat: jnp.ndarray
+  ) -> Tuple[jnp.ndarray, jnp.ndarray, float, float]:
+
+    (
+        self.state_eta, self.state_xi, eta_predictions, xi_predictions, loss_a,
+        loss_b
+    ) = self.unbalancedness_handler.step_fn(
+        source=source,
+        target=target,
+        condition=source_conditions,
+        a=tmat.sum(axis=1),
+        b=tmat.sum(axis=0),
+        state_eta=self.unbalancedness_handler.state_eta,
+        state_xi=self.unbalancedness_handler.state_xi,
+    )
+    return eta_predictions, xi_predictions, float(loss_a), float(loss_b)
+
+
+class GENOTLin(GENOTBase):
+  """Implementation of GENOT-L (:cite:`klein:23`).
+
+  GENOT-L (Generative Entropic Neural Optimal Transport, linear) is a
+  neural solver for entropic (linear) OT problems.
+  """
+
+  def __call__(self, train_loader, valid_loader):
```

Comment:
Is this the correct signature?  I thought this had changed to four loaders?

Reply:
I think you're right, it should be four loader. But for now, @michalk8 wanted to do the final clean, right?